Deconv

wangyang59 opened the pull request:
Implement CPU version of convolution transpose (deconv) layer using expand method.

qingqing01 commented on paddle/gserver/layers/ExpandConvTransLayer.cpp:
```diff
@@ -0,0 +1,337 @@
+      expandInData += subK * subN;
+    }
+  }
+}
```
1. Actually, the forward of de-conv is the backward of convolution layer. And the backward is the forward of convolution layer. So it's better to reuse the implementation in convolution layer.  This can refer the implementation method in Caffe.
2.  The name DeConvLayer may be better.

qingqing01 replied:
I see. It's  better to refactor code. It might be able to extract the same code into forwardWeight/backwardWeight  and forwardBias/backwardBias (or other suitable name), then call these functions in ExpandConvLayer and ExpandConvTransLayer.

wangyang59 commented:
Hi @qingqing01 
I have refactored the code following your suggestions and now ExpandConvLayer and ExpandConvTransLayer share a common superclass of ConvBaseLayerCpu which contains most of the shared codes. 
Could you please help take a look at it again? Thanks!

luotao1 commented on paddle/gserver/layers/ConvBaseLayer.h:
```diff
@@ -27,6 +27,9 @@ class ConvBaseLayer : public Layer {
   typedef std::vector<int> IntV;
 
+  /// True if it's convolution layer, false if it's deconv layer
+  bool isConv_;
```
变量名改成isDeConv，因为都是ConvLayer

wangyang59 replied:
done

luotao1 commented on paddle/gserver/layers/ConvBaseLayer.h:
```diff
@@ -106,6 +111,18 @@ class ConvBaseLayer : public Layer {
+         (outputSize - 1) * stride + filterSize - 2 * padding - stride + 1;
+    } else {
+     imageSize = (outputSize - 1) * stride + filterSize - 2 * padding;
+    }
```
```
int imageSize = (outputSize - 1) * stride + filterSize - 2 * padding;
if (!caffeMode_) {
  imageSize += - stride + 1;
}
```

luotao1 replied:
可以直接调用mathUtils.h中的outputSize函数（就多了最后一个caffeMode_的变量），不用在这里重新写一份。另外，把imageSize函数也放到mathUtils.h里吧

luotao1 commented on paddle/gserver/layers/ConvBaseLayerCpu.h:
```diff
@@ -0,0 +1,91 @@
+  /// The spatial dimensions of height of output feature map.
+  IntV outputH_;
+  /// The spatial dimensions of width of output feature map.
+  IntV outputW_;
```
imgSizeH_, imgSizeW_, outputH_, outputW_这四个变量在ConvBaseLayer里面有，可以直接继承，不用再写一遍了

wangyang59 replied:
Sorry, I didn't realize that there was a change by @qingqing01 made to ConvBaseLayer and ExpandConvLayer

luotao1 commented on paddle/gserver/layers/ConvBaseLayerCpu.h:
```diff
@@ -0,0 +1,91 @@
+/**
+ * @brief A subclass of ConvBaseLayer that is a superclass of both
+ * ExpandConvLayer and ExpandConvTransLayer
+ */
```
既然是ExpandConvLayer and ExpandConvTransLayer的superclass，名字不要叫ConvBaseLayerCpu。可以叫ExpandConvBaseLayer，或者别的，名字要有所反应。

wangyang59 replied:
I have renamed it to ExpandConvBaseLayer

luotao1 commented on paddle/gserver/layers/ConvBaseLayerCpu.h:
```diff
@@ -0,0 +1,91 @@
+
+  /*The expandInput_ and transOutValue_ are used for CPU expand conv calc*/
+  /// Expand one sample at a time. shape:
+  /// (numChannels * filterPixels_, outputSizeH * outputSizeW)
```
46-48行的注释，要不全用///，要不就用/**/

wangyang59 replied:
done

luotao1 commented on paddle/gserver/layers/ExpandConvLayer.cpp:
```diff
@@ -166,7 +70,7 @@ void ExpandConvLayer::forward(PassType passType) {
     for (size_t off = 0; off < image->getHeight(); off++) {
       REGISTER_TIMER_INFO("expandFwdOnce", getName().c_str());
-      expandFwdOnce(image, i, off);
+      expandFwdOnce(image, getOutputValue(), i, off);
```
在for循环外：MatrixPtr outV = getOutputValue(); for循环里面直接调用outV，可以少访问image->getHeight()次指针。

wangyang59 replied:
done

luotao1 commented on paddle/gserver/layers/ExpandConvLayer.cpp:
```diff
@@ -218,109 +98,16 @@ void ExpandConvLayer::backward(const UpdateCallback &callback) {
   for (size_t i = 0; i != inputLayers_.size(); ++i) {
     /* First, calculate the input layers error */
-    bpropActs(outGrad, i);
+    if (NULL != getPrev(i)->getOutputGrad()) {
```
if语句改成：if (getPrev(i)->getOutputGrad())

wangyang59 replied:
done

luotao1 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1067,6 +1067,37 @@ def parse_conv(conv, input_layer_name, conv_conf):
             - conv.filter_size) / float(conv.stride)))
 
+
+def parse_convt(conv, input_layer_name, conv_conf, num_filters):
```
改成parse_conv_trans比较容易懂，实现的时候，1072-1098行有些可以复用parse_conv代码

wangyang59 replied:
done

luotao1 commented on paddle/gserver/tests/test_LayerGrad.cpp:
```diff
@@ -262,6 +262,8 @@ void testConvLayer(const string& type, bool trans, bool useGpu) {
 
   testLayerGrad(config, "conv", 100, trans, useGpu);
+  // Use small batch_size and useWeight=true to test biasGrad
+  testLayerGrad(config, "conv", 2, trans, useGpu, true, 0.02);
```
batchsize=2也太小了，10或者100

wangyang59 replied:
The reason I set the batchsize to 2 is that the bug in calculating the bpropSharedBias cannot be caught with large batchsize due to averaging out errors.

luotao1 replied:
请问有什么bug？可以贴出来么

wangyang59 replied:
void ExpandConvBaseLayer::bpropSharedBias(MatrixPtr biases, MatrixPtr v) {
  size_t mapW = getOutputSize() / numFilters_;
  size_t mapH = v->getElementCnt() / mapW;
  MatrixPtr vTmp = Matrix::create(v->getData(), mapH, mapW, false, useGpu_);

  Matrix::resizeOrCreate(transOutValue_, mapW, mapH, false, useGpu_);

  vTmp->transpose(transOutValue_, false);  // false means no memory allocation
  transOutValue_->reshape(transOutValue_->getElementCnt() / numFilters_,
                          numFilters_);
  biases->collectBias(*transOutValue_, 1.0f);
}

The last two lines was 
  vTmp->reshape(transOutValue_->getElementCnt() / numFilters_,
                          numFilters_);
  biases->collectBias(*vTmp, 1.0f);

luotao1 replied:
你的意思是改过后，就不会出bug？那可以修正下吧

wangyang59 commented on paddle/gserver/layers/ExpandConvTransLayer.cpp:
```diff
@@ -0,0 +1,122 @@
+  resetOutput(batchSize, getSize());
+
+  MatrixPtr output = nullptr;
+  for (size_t i = 0; i != inputLayers_.size(); ++i) {
```
done

wangyang59 commented on python/paddle/trainer_config_helpers/layers.py:
```diff
@@ -1625,6 +1626,128 @@ def img_conv_layer(input, filter_size, num_filters,
+@wrap_bias_attr_default()
+@wrap_act_default(act=ReluActivation())
+@layer_support(DROPOUT)
+def img_convTrans_layer(input, filter_size, num_filters,
```
done

wangyang59 commented:
Hi @luotao1 I have modified my codes per your suggestions. 

@qingqing01 , when I rebase my code, I realized that you have also modified ConvBaseLayer and ExpandConvBaseLayer at the same time (pull#218). I have merged my changes with yours,  so could you please also have a look at it to make sure that I didn't break your code? Thanks!

luotao1 commented:
Note that 'All checks have failed', you can click 'Details' to see the log.

luotao1 commented on paddle/gserver/layers/ConvBaseLayer.h:
```diff
@@ -74,6 +77,8 @@ class ConvBaseLayer : public Layer {
   bool caffeMode_;
 
+
+
```
多余的两行空行可以去掉

luotao1 commented on paddle/gserver/layers/ExpandConvBaseLayer.cpp:
```diff
@@ -0,0 +1,263 @@
+   * convTrans, and in other functions too.
+   * */
+  int channel;
+  int nf;
```
nf变量改成numFilters，比如直接

wangyang59 replied:
done

luotao1 commented on paddle/gserver/layers/ExpandConvBaseLayer.cpp:
```diff
@@ -0,0 +1,263 @@
+  /* Initialize the projection */
+  for (auto &inputConfig : config_.inputs()) {
+    const ConvConfig &conf = inputConfig.conv_conf();
+    nf = (!isDeconv_) ? numFilters_ : conf.channels();
```
nf = isDeconv_ ? conf.channels() : numFilters_; 41行也同样

wangyang59 replied:
done

luotao1 commented on python/paddle/trainer_config_helpers/layers.py:
```diff
@@ -1619,10 +1636,10 @@ def img_conv_layer(input, filter_size, num_filters,
         bias=ParamAttr.to_bias(bias_attr),
         shared_biases=shared_biases,
-        type=LayerType.CONV_LAYER,
+        type=lt,
```
type = LayerType.CONVTRANS_LAYER if trans else LayerType.CONV_LAYER, 去掉1622-1625行

luotao1 commented on python/paddle/trainer_config_helpers/layers.py:
```diff
@@ -1607,6 +1618,12 @@ def img_conv_layer(input, filter_size, num_filters,
+    if trans:
+        lt = LayerType.CONVTRANS_LAYER
+    else:
+        lt = LayerType.CONV_LAYER
```
lt = LayerType.CONVTRANS_LAYER if trans else LayerType.CONV_LAYER

luotao1 replied:
lt = LayerType.CONVTRANS_LAYER if trans else LayerType.CONV_LAYER

wangyang59 replied:
done

luotao1 commented on python/paddle/trainer_config_helpers/layers.py:
```diff
@@ -36,7 +36,7 @@
            "cos_sim", "hsigmoid", "conv_projection",
            "regression_cost", 'classification_cost', "LayerOutput",
-           'img_conv_layer', 'img_pool_layer', 'batch_norm_layer',
+           'img_conv_layer', 'img_convTrans_layer', 'img_pool_layer', 'batch_norm_layer',
```
img_convTrans_layer去掉，不然导致单测不过

wangyang59 replied:
done

luotao1 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1105,6 +1105,37 @@ def parse_conv(conv, input_layer_name, conv_conf):
             - conv.filter_size) / float(conv.stride)))
 
+
+def parse_conv_trans(conv, input_layer_name, conv_conf, num_filters):
```
parse_conv_trans和parse_conv的共用代码很多，能否在parse_conv里面加一个trans变量。同样ConvTransLayerBase和ConvLayerBase的共用代码也很多，能否在ConvLayerBase里面加一个trans变量。

wangyang59 replied:
done for the parse_conv_trans
However for the ConvTransLayerBase, I am not sure I can do the same thing since it has a decorator and the overlapping of the codes is not that big.

luotao1 commented on paddle/gserver/tests/test_LayerGrad.cpp:
```diff
@@ -311,6 +313,51 @@ TEST(Layer, convLayer) {
+
+TEST(Layer, convTransLayer) {
+  testConvTransLayer("exconvt", /* trans= */ false, /* useGpu= */ false);
+/*
```
为什么不能调用testConvLayer呢

luotao1 commented on paddle/gserver/tests/test_ConvTrans.cpp:
```diff
@@ -0,0 +1,251 @@
+    result = Matrix::create(1, imgSize * imgSize, false, false);
+    result->zeroMem();
+    result->add(1.0);
+    doOneConvtTest(imgSize, output_x, stride, padding, filter_size, result);
```
去掉195-199行，写成这种类型：doOneConvtTest(/* imgSize */ 5, /* output_x */ 1, /* stride */ 1, /* padding */ 0, /* filter_size */ 5, result); 下同

wangyang59 replied:
done

luotao1 commented on paddle/gserver/tests/test_ConvTrans.cpp:
```diff
@@ -0,0 +1,251 @@
+
+TEST(Layer, convTransLayerFwd2) {
+    size_t imgSize, output_x, stride, padding, filter_size;
+    MatrixPtr result;
```
200行出现多次，可在193行就进行初始化：MatrixPtr result = Matrix::create(1, imgSize * imgSize, false, false);

wangyang59 replied:
done

luotao1 commented:
目前可以在GPU上运行么？

wangyang59 commented on paddle/gserver/layers/ConvBaseLayer.cpp:
```diff
@@ -71,23 +89,45 @@ size_t ConvBaseLayer::calOutputSize() {
+    }
+    getOutput().setFrameHeight(imgSizeH_[0]);
+    getOutput().setFrameWidth(imgSizeW_[0]);
+    layerSize = imgSizeH_[0] * imgSizeW_[0] * size_t(numFilters_);
```
done

wangyang59 commented:
@luotao1 Thank you very much for your detailed review and comments!
I am new to Paddle and C++ and a larger engineering project, so I really appreciate you for helping me improve my code quality here :)
I have changed my code per your suggestions. Please have a look at it again.

luotao1 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1613,6 +1631,73 @@ class ConvLayer(ConvLayerBase):
+#            (parallel_nn == 0 or self.config.device > -1)):
+#             self.layer_type = "cudnn_convt"
+#         else:
+#             self.layer_type = "exconvt"
```
1661-1665注释掉的行可以都去掉

luotao1 commented on paddle/gserver/tests/test_LayerGrad.cpp:
```diff
@@ -311,6 +313,49 @@ TEST(Layer, convLayer) {
+  testConvTransLayer("exconvt", /* trans= */ false, /* useGpu= */ false);
+#ifndef PADDLE_ONLY_CPU
+  testConvTransLayer("exconvt", /* trans= */ false, /* useGpu= */ true);
+  // testConvLayer("cudnn_conv", /* trans= */ false, /* useGpu= */ true);
```
355行为什么注释掉呢

wangyang59 replied:
because I have not implemented cudnn_convt yet

wangyang59 replied:
I have removed that 355. However, line 354 has to be inside the condition of 353.

luotao1 commented on paddle/gserver/layers/ConvBaseLayer.cpp:
```diff
@@ -71,23 +89,44 @@ size_t ConvBaseLayer::calOutputSize() {
+         outW.push_back(
+             outputSize(inW[i], filterSize_[i], padding_[i], stride_[i]));
+         CHECK_EQ(outH[i], outH[0]);
+         CHECK_EQ(outW[i], outW[0]);
```
115-116行这两个check应该在if else外面

wangyang59 replied:
done

wangyang59 commented on paddle/gserver/layers/ConvBaseLayer.cpp:
```diff
@@ -22,6 +22,12 @@ bool ConvBaseLayer::init(const LayerMap& layerMap,
+    isDeconv_ = false;
+  } else {
+    isDeconv_ = true;
+  }
```
done

wangyang59 commented:
@luotao1 I have modified the code according to your requests.

luotao1 commented:
请看下test_LayerGrad中batchsize=2的问题，是否是bug？

wangyang59 commented:
@luotao1 I have modified the code accordingly. 
As I explained earlier, the batch_size was set to 2 in test_LayerGrad because only small batch size can catch the error in calculating bias gradient, otherwise the error would be too small to detect due to averaging out.

wangyang59 commented:
@luotao1 I have merged with your pull request "abstract outputSize function in CNN-related layers (#314)"
Please have a look to see if my merge makes sense to you.

luotao1 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1093,18 +1097,41 @@ def parse_conv(conv, input_layer_name, conv_conf):
+        else:
+            conv_conf.output_x = \
+                1 + int(math.ceil((2 * conv.padding + conv_conf.img_size \
+                - conv.filter_size) / float(conv.stride)))
```
python里面计算outputSize的可以调用cnn_output_size函数

wangyang59 replied:
done

luotao1 commented on paddle/gserver/tests/test_LayerGrad.cpp:
```diff
@@ -312,6 +314,48 @@ TEST(Layer, convLayer) {
+  conv->set_output_x(
+      (2 * conv->padding() + conv->img_size() - conv->filter_size()) /
+          ((float)conv->stride()) +
+      1.5);
```
单测里面也可以直接调用outputSize函数

luotao1 commented:
@qingqing01 关于batchsize=2的问题，你怎么看？@wangyang59 你说的error无法检测到，是单测写的问题，还是layer中的问题。因为图像的batchSize在训练和预测中一直很大，如果是单测写法问题，就修改单测的，如果是layer中的问题，就修改layer的。

wangyang59 commented on paddle/gserver/tests/test_ConvTrans.cpp:
```diff
@@ -0,0 +1,248 @@
+    conv->set_output_x(
+        (2 * conv->padding() + conv->img_size() - conv->filter_size()) /
+            ((float)conv->stride()) +
+        1.5);
```
done

wangyang59 commented on paddle/gserver/tests/test_LayerGrad.cpp:
```diff
@@ -312,6 +314,48 @@ TEST(Layer, convLayer) {
+#ifndef PADDLE_ONLY_CPU
+  testConvTransLayer("exconvt", /* trans= */ false, /* useGpu= */ true);
+#endif
+}
```
done

wangyang59 commented:
@luotao1 Thanks for your comments. I have modified the code according to your suggestions. 
Can we have a meeting to discuss remaining issues at 1pm Beijing time today (Wednesday)?

qingqing01 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1612,6 +1635,68 @@ class ConvLayer(ConvLayerBase):
+        # if set type=conv, but still reserve the way user specify
+        # exconv or cudnn_conv manually.
+        if self.layer_type == "cudnn_convt":
+            config_assert(use_gpu, "cudnn_convt only support GPU")
```
If not support cuDNN,  the lines 1656 - 1663 can be removed.

wangyang59 replied:
done

qingqing01 commented on paddle/gserver/layers/ExpandConvTransLayer.h:
```diff
@@ -0,0 +1,44 @@
+ * This layer expands input and use matrix multiplication to
+ * calculate convolution transpose (deconv) operation.
+ *
+ * The config file api is img_convTrans_layer.
```
The config file api is not `img_convTrans_layer`.

wangyang59 replied:
done

qingqing01 commented on paddle/gserver/layers/ExpandConvTransLayer.cpp:
```diff
@@ -0,0 +1,94 @@
+
+  /* malloc memory for the output_ if necessary */
+  /* note: one sample correspond to one colum, and the
+   *   transOutValue correspond sample to one row */
```
I'm not quite sure I understand the note.

luotao1 commented:
Sure, we can have a meeting to discuss remaining issues at 1pm Beijing time today (Wednesday).

luotao1 commented on python/paddle/trainer/config_parser.py:
```diff
@@ -1093,18 +1097,37 @@ def parse_conv(conv, input_layer_name, conv_conf):
+        else:
+            conv_conf.img_size = \
+                (conv_conf.output_x - 2) * conv.stride \
+                + conv.filter_size - 2 * conv.padding + 1 
```
1123-1130可以写个cnn_image_size函数

wangyang59 replied:
done

luotao1 commented:
缺少python文件的单测，可参考https://github.com/baidu/Paddle/tree/develop/python/paddle/trainer_config_helpers/tests, 在configs目录下修改或添加新的单测，在protostr目录下，修改或添加新生成的protostr

luotao1 commented:
#263

wangyang59 commented:
@luotao1 @qingqing01 The build on mac is not stable. Build #724 failed and build #725 passed although they have identical code base.
